[DMS-39] Support labels, breaks, and continues

DK318 opened the pull request:
`break <label>` and `continue <label>` statements are parsed into the `BreakE` node, so, supporting breaks means that we support continues.

## `LabelE` as a Statement
The `LabelE` node could appear only next to the loops. `motoko`'s parser additionally wraps the loop's body with the label `continue <label_name>`. The final translation looks like this.

Motoko's pseudocode:
```
label while_loop while (cond) {
  ...
};
```

Parsed code:
```
(LabelE while_loop
  (WhileE cond (LabelE "continue while_loop" exp))
)
```

Translated code:
```
while (cond) {
 ...
 label $continue_while_loop
}
label $while_loop
```

`break`s in this case are translated into `goto`s.

## `LabelE` as an Assignment RHS
`let v = label lbl : <typ> <exp>` case is translated to
```
var $t_lbl: <typ>;
<tr_exp>;
label $lbl;
v := $t_lbl;
```
where `<tr_expr>` is the translated expression with one hack: the last declaration in the `BlockE` node is wrapped with `BreakE`. `break`s in this case are handled like `return`s:
```
break lbl(<exp>) => { $t_lbl := <exp>; goto $lbl; }
```

int-index commented on src/viper/trans.ml:
```diff
@@ -766,6 +830,22 @@ and rets t_opt =
+          if Char.equal ch ' '
+          then '_'
+          else ch)
+      |> Format.asprintf "$%s"
```
We use `$` to avoid conflicts between compiler-generated and user-written identifiers. Since the user chooses the name for the label, adding `$` in front of it does not fit this scheme.

Also I wonder if Motoko has different namespaces for labels and variables. That is, is it legal to have `label x` and `let x` in scope at the same time?

DK318 replied:
In motoko labels and variables have different namespaces. However, in viper they share the same one. 
Right now I see 2 solutions:
1. Prefix it with `$lbl_`.
2. Handle it with `fresh_id` somehow. 

The latter sounds better but we should guarantee that `fresh_id` was called at least once for all `decl`s. 

Btw do we care about counters after variable names? Can't we just strip stamps map and use the global counter instead?

int-index replied:
Given a label `x`, I suggest we generate
1. `$lbl$break$x`
2. `$lbl$continue$x`

Or, if it's easier, `$lbl$x` and `$lbl$continue$x`

DK318 replied:
> Or, if it's easier, $lbl$x and $lbl$continue$x

Supported this variant

int-index commented on src/viper/syntax.ml:
```diff
@@ -103,4 +103,3 @@ and typ' =
   | TupleT
   | OptionT of typ
   | ConT of id * typ list
-
```
Please avoid unnecessary formatting changes (especially the trailing newline, I think we were asked to keep it)

DK318 replied:
Just out of curiosity. Shall we keep 2 trailing newlines at the end of the file?

int-index commented on src/viper/trans.ml:
```diff
@@ -859,3 +939,33 @@ and tuple_alloc at ctxt lhs ts es : stmt list =
+  let label_rhs =
+    match label_rhs.it with
+    | M.BlockE decs ->
+      let rec map_last ~f = function
```
This is a general-purpose helper, so please move it to the top level

int-index commented on src/viper/trans.ml:
```diff
@@ -100,6 +100,8 @@ module Env = T.Env
   { self : string option;
     ids : (sort * T.t) T.Env.t;
+    label_vars : (string, id) Hashtbl.t; (* Orig label var -> associated tmp variable. *)
+    mapped_labels : (string, string) Hashtbl.t; (* Orig label var -> label var in viper. *)
```
If we make the translation of label names use a deterministic scheme (like the one I suggest in https://github.com/serokell/motoko/pull/18/files#r1621415198), then we don't need this hashmap, do we? Just have a toplevel `string -> string` function.

As I understand it, the current translation makes use of `fresh_id`, hence nondeterminism and the need to track those labels in the context. But it's not clear to me what problem we are solving with this use of `fresh_id`. Does it help us to avoid conflicts? If yes, could you demonstrate when such conflicts arise, is there an example?

DK318 replied:
> But it's not clear to me what problem we are solving with this use of fresh_id

Labels in `viper` couldn't be shadowed, so, we need to use `fresh_id` here.

> could you demonstrate when such conflicts arise, is there an example?

Yeah, `label-break-continue.mo` already has such example. Let me point it out:
```
    let block_label_early_expr = label block : (Int, Int) {
      if (true) break block(42, 42);
      (24, 24)
    };

    let block_label_expr = label block : (Int, Int) {
      if (false) break block(42, 42);
      (24, 24)
    };
```
Here we want to distinguish these two `block` labels

int-index replied:
OK, I see, thank you for the explanation. I also see that it is covered by a test case, good.

DK318 commented on src/viper/trans.ml:
```diff
@@ -100,6 +100,8 @@ module Env = T.Env
 type ctxt =
   { self : string option;
     ids : (sort * T.t) T.Env.t;
+    label_vars : (string, id) Hashtbl.t; (* Orig label var -> associated tmp variable. *)
```
> Does it have to be a mutable hashtable

Yeah, `Hashtbl.t` in OCaml is mutable.

> could we use an immutable map?

Changed them to immutable maps

GoPavel reviewed (changes requested):
Okay I would like to see clear separation between loop labels and labeled expressions.
I am a bit worried about the latter part. It looks likes overcomplication of translator. TBH I didn't expect you to implement it.

Problem is that `goto` seems like is not stable feature of Viper. There is not so much documentation about it and I think `goto` was added for inferring loops mostly. I would use it as less as possible.
I have spent some time to find issue in Carbon, Silver or Silicon repos but I didn't. Let's keep this but make separation clear to isolate this part.

GoPavel commented:
However, labeled expressions could be good step to `do { }` expressions that is much more common I think

DK318 commented on src/viper/trans.ml:
```diff
@@ -766,6 +832,22 @@ and rets t_opt =
+and loop_label ctxt loop_label =
+  let label_name =
+    loop_label.it
+      |> String.map (fun ch ->
+          (* Life without Core is hard... *)
+          if Char.equal ch ' '
+          then '$'
+          else ch)
```
> I don't think that Motoko label can contain spaces.

`continue` statements are parsed into `BreakE` with `"continue " ^ label` name ([reference](https://github.com/serokell/motoko/blob/master/src/mo_frontend/parser.mly#L703)). This part handles that.

GoPavel replied:
okay, fair enough
But I would replace it with a code doing exact that:
```suggestion
 let label_name = match String.split_on_char ' ' loop_label.it with
                 | ["continue"; name] -> "continue$" ^ name
                 | _                  -> loop_label.it 
 in 
 let label_name = fresh_id ("$lbl$" ^ label_name)
```

DK318 replied:
Applied

DK318 commented on test/viper/label-break-continue.mo:
```diff
@@ -0,0 +1,70 @@
+actor LabelBreakContinue {
+  func label_expressions() {
+    let simple_label = label simple : Int break simple(42);
+    assert simple_label == 42;
```
Done

DK318 commented on src/viper/trans.ml:
```diff
@@ -859,3 +941,29 @@ and tuple_alloc at ctxt lhs ts es : stmt list =
 and tuple_prj ctxt at e1 e2 t =
   prjE at (exp ctxt e1) (exp ctxt e2) (typed_field t)
 
+and label_alloc ~label_id ~label_type ~label_rhs ~label_note at ctxt lhs : seqn' =
```
Done

DK318 commented:
> Let's keep this but make separation clear to isolate this part.

Not sure that I understand this part. The translation of label expression is handled by `label_expr_alloc`. The second part is handled in `M.BreakE` pattern matching (the one about variable assignment). I extracted it to a separate function (`compile_break`).

GoPavel commented:
> > Let's keep this but make separation clear to isolate this part.
> 
> Not sure that I understand this part. The translation of label expression is handled by `label_expr_alloc`. The second part is handled in `M.BreakE` pattern matching (the one about variable assignment). I extracted it to a separate function (`compile_break`).

I meant it is good to see clear separation of 2 features: expression labels and `continue`/`break`.
Now I see that for `continue`/`break` we have 3 places where code is changed:
* a map `loop_labels` in the state
* match on `BreakE`  and `LabelE` expressions in the `stmt` function
* compiling while loop

Meanwhile for expression labels it is (almost) different places:
* `label_expr_alloc` in `assign_stmts`
* and in a match on `BreakE` again.
Right..? after code investigation I understood that it's true.

A couple of questions:
* Why `compile_break` compiles only expression inside break if it's expr. label? From the name I would suppose that it is compiling whole `BreakE`. Suggestion: move body of the `compile_break` back to the match branch.
* After diving a bit more into code I have realized that `loop_labels` actually is not loop labels but rather `break`/`continue` labels. Suggestion: rename the map or just move all labels into one with kind field `Break | Continue | Custom of str` or smth like that 
* I see that `BreakE` could be in label expression as well as in loop statement. This two situation now in code is very very vague. The code is doing following:
   *  check the `expr_labels` map: if a label exists then compile an assignment, otherwise compile expr (_that is always empty tuple in the loop_)
   * check the `loop_labels` map: if a label doesn't exists fail (_even if it's a label expression situation? Guess it's should be a dead code then since tests pass_).
  * emit `goto`

  I would suggest:
  * Extract code about loop situation into separate branch like `BreakE {it=TupE []}`. Okay, then we drop label expression of unit type, but it's very rare and not interesting for us moment IMHO.
 
Sorry, maybe I was very unclear in my previous review. At first, I am trying to keep code clear and as simple as possible. And the second pointer here that I want to have opportunity to comment out pars of code related label expressions because I don't feel confident in Viper support of `goto`. Now I see it's very hard to do.

DK318 commented:
I believe that I've addressed your comments.

> Why compile_break compiles only expression inside break if it's expr. label? From the name I would suppose that it is compiling whole BreakE. Suggestion: move body of the compile_break back to the match branch.

Stripped that.

> After diving a bit more into code I have realized that loop_labels actually is not loop labels but rather break/continue labels. Suggestion: rename the map or just move all labels into one with kind field Break | Continue | Custom of str or smth like that

Yeah, naming is a bit strange. I refreshed my knowledge of this code and did the next renamings:
1. `loop_labels` -> `label_to_vpr_label`
2. `expr_labels` -> `label_to_tmp_var`

> I see that BreakE could be in label expression as well as in loop statement. This two situation now in code is very very vague.

Applied your suggestion.

Btw I added a comment to `label_expr_alloc`. I believe it should be clearer what this function does.

GoPavel commented on src/viper/trans.ml:
```diff
@@ -859,3 +938,44 @@ and tuple_alloc at ctxt lhs ts es : stmt list =
   prjE at (exp ctxt e1) (exp ctxt e2) (typed_field t)
 
+and label_expr_alloc ~label_id ~label_type ~label_rhs ~label_note at ctxt lhs : seqn' =
+  (* Here is a hack: instead of assigning a unit expression
+     we'll just treat them as statements. The next Motoko's code
+
+     {[
+      let x = label unit_lbl : () { break unit_lbl(v := 42); unreachable_expr; };
+     ]}
+
+     would be translated to
+
+     {[
+      v := 42;
+      goto unit_lbl;
+      unreachable_expr;
+      label unit_lbl;
+     ]} *)
```
Yes, but this comment mostly is about last line. While whole code after is about adding explicit `break` to the last expression which is returning from the block.
Add comment about this as well, maybe it would be nicer if you show example of Motoko to Motoko translation.

DK318 replied:
Moved this comment and added an extra one

GoPavel commented:
Okay, fix comment above and you can merge it
